lib/csv/decoder.ex

```diff
@@ -61,120 +61,102 @@ defmodule CSV.Decoder do
-  end
-  defp lex_to_parse(result) do
-    result
+  defp decode_row({ nil, 0 }, %{ row_length: false }), do: { :ok, [] }
+  defp decode_row({ line, index }, %{ headers: headers, row_length: row_length } = options) do
+    with { :ok, parsed, _ } <- parse_row({ line, index }, options |> Enum.into([])),
+         { :ok } <- check_row_length({ parsed, index }, row_length),
+         do: build_row(parsed, headers)
   end
 
-  defp check_row_length({ :ok, data, index }, row_length) do
-    actual_length = data |> Enum.count
-
-    case actual_length do
-      ^row_length -> { :ok, data, index }
-      _ -> { :error, RowLengthError, "Encountered a row with length #{actual_length} instead of #{row_length}", index }
-    end
-  end
-  defp check_row_length(error, _) do
-    error
+  defp parse_row({ line, index}, options) do
+    with { :ok, lex, _ } <- Lexer.lex({ line, index }, options),
+    do: Parser.parse({ lex, index }, options)
   end
 
-  defp build_row({ :ok, data, _ }, headers) when is_list(headers) do
-    headers |> Enum.zip(data) |> Enum.into(%{})
+  defp aggregate(stream, true) do
+    stream |> LineAggregator.aggregate
   end
-  defp build_row({ :ok, data, _ }, _) do
-    data
+  defp aggregate(stream, false), do: stream
+
+  defp build_row(data, headers) when is_list(headers) do
+    { :ok, headers |> Enum.zip(data) |> Enum.into(%{}) }
   end
-  defp build_row(error, _) do
-    error
+  defp build_row(data, _), do: { :ok, data }
+
+  defp options_with_defaults(options) do
+    num_pipes = options |> Keyword.get(:num_pipes, Defaults.num_workers)
+    options
+    |> Keyword.merge(num_pipes: num_pipes,
+                     num_workers: options |> Keyword.get(:num_workers, num_pipes),
+                     multiline_escape: options |> Keyword.get(:multiline_escape, true),
+                     headers: options |> Keyword.get(:headers, false),
+                     row_length: false)
+    |> Enum.into(%{})
   end
 
-  defp get_headers(headers, stream, _) when is_list(headers) do
-    { headers, stream }
+  defp prepare_headers({ stream, %{ headers: true } = options }) do
+    { stream |> Stream.drop(1),
+      %{ options | :headers => get_first_row(stream, options) } }
   end
-  defp get_headers(headers, stream, options) when headers do
-    headers = stream
-              |> get_first_row(options)
+  defp prepare_headers(job), do: job
 
-    { headers, stream |> Stream.drop(1) }
+  defp prepare_row_length({ stream, %{ headers: headers } = options}) do
+    { stream, %{ options | :row_length => get_row_length(headers || stream, options) } }
   end
-  defp get_headers(_, stream, _) do
-    { false, stream }
+
+  defp check_row_length(_, false), do: { :ok }
+  defp check_row_length({ data, index }, row_length) do
+    actual_length = get_row_length(data)
+
+    case actual_length do
+      ^row_length -> { :ok }
+      _ -> { :error, RowLengthError, "Encountered a row with length #{actual_length} instead of #{row_length}", index }
```

Comment:
I don't see the downstream issue, although I see it could be an issue if we change `with` back to `case` here https://github.com/beatrichartz/csv/pull/33/files#diff-9609334881ddb4c8ef0b91112bfa9991R87.

Regardless, I think your idea for a `build_row_with_length` function makes more sense than the current, so I'll go for it.